ZombieProcess errors in tests

psutil may raise `psutil.ZombieProcess`, which has happened in our tests occasionally,

```bash
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================ tests coverage ================================
_______________ coverage: platform linux, python 3.11.11-final-0 _______________

Coverage XML written to file test/coverage.xml
=========================== short test summary info ============================
FAILED test/integration/test_run_command_integration.py::test_run_interrupted[Success] - psutil.ZombieProcess: PID still exists but it's a zombie (pid=4646)
====== 1 failed, 38 passed, 3 skipped, 144 warnings in 116.28s (0:01:56) =======
Error: Process completed with exit code 1.
```

![Image](https://github.com/user-attachments/assets/730040b5-b8f0-47cb-864d-8132030f57d5)

https://github.com/giampaolo/psutil/blob/5a605b2e0bfaa724454327aa553fb2a30cc0a8a7/psutil/_pslinux.py#L1696-L1698

--- a/autosubmit/helpers/processes.py
+++ b/autosubmit/helpers/processes.py
@@ -13,11 +13,12 @@
 # GNU General Public License for more details.
 
 import re
+from contextlib import suppress
 from getpass import getuser
 from pathlib import Path
 from typing import List, Optional
 
-from psutil import process_iter
+from psutil import process_iter, ZombieProcess
 
 from log.log import Log
 
@@ -143,15 +144,21 @@ def retrieve_expids() -> List[str]:
     """
     user: str = getuser()
     expids: List[str] = []
-    for process in [p for p in process_iter(['pid', 'cmdline', 'username']) if p.username() == user]:
-        expid = _match_autosubmit_cmdline(process.cmdline())
-        if expid:
-            expids.append(expid)
+    # NOTE: psutil may raise ``ZombieProcess`` in some cases, even when other exceptions happen in the
+    #       case when a process is a zombie (e.g. got a IO error, but the process is zombie)
+    with suppress(ZombieProcess):
+        for process in [p for p in process_iter(['pid', 'cmdline', 'username']) if p.username() == user]:
+            with suppress(ZombieProcess):
+                expid = _match_autosubmit_cmdline(process.cmdline())
+                if expid:
+                    expids.append(expid)
 
     return expids
 
 
-# TODO: check with Dani if the platform is needed (no processes using it on hub or destine vm?)
+# TODO: check with Dani if the platform is needed (no processes using it on hub or destine vm?).
+#       The previous code used ``grep`` to search by expid, and also had some code ready to
+#       search by platform name, although that was not actually used anywhere, yet.
 def process_id(expid: str, command="run") -> Optional[int]:
     """Retrieve the process id of the autosubmit process.
 
@@ -163,11 +170,16 @@ def process_id(expid: str, command="run") -> Optional[int]:
     :rtype: Optional[int]
     """
     user: str = getuser()
-    processes = [
-        p for p
-        in process_iter(['pid', 'cmdline', 'username'])
-        if p.username() == user and _match_autosubmit_cmdline(p.cmdline(), command, expid)
-    ]
+    try:
+        processes = [
+            p for p
+            in process_iter(['pid', 'cmdline', 'username'])
+            if p.username() == user and _match_autosubmit_cmdline(p.cmdline(), command, expid)
+        ]
+    except ZombieProcess:
+        # NOTE: psutil may raise ``ZombieProcess`` in some cases, even when other exceptions happen in the
+        #       case when a process is a zombie (e.g. got a IO error, but the process is zombie)
+        return None
 
     if not processes:
         return None

--- a/autosubmit/profiler/profiler.py
+++ b/autosubmit/profiler/profiler.py
@@ -1,19 +1,17 @@
-#!/usr/bin/env python3
-
-# Copyright 2015-2024 Earth Sciences Department, BSC-CNS
-
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
 # This file is part of Autosubmit.
-
+#
 # Autosubmit is free software: you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
 # the Free Software Foundation, either version 3 of the License, or
 # (at your option) any later version.
-
+#
 # Autosubmit is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU General Public License for more details.
-
+#
 # You should have received a copy of the GNU General Public License
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
 
@@ -21,28 +19,29 @@ import cProfile
 import io
 import os
 import pstats
-from enum import Enum
 from datetime import datetime
+from enum import Enum
 from pathlib import Path
 from pstats import SortKey
 
-import psutil
-from autosubmitconfigparser.config.basicconfig import BasicConfig
+from psutil import Process
 
+from autosubmitconfigparser.config.basicconfig import BasicConfig
 from log.log import Log, AutosubmitCritical
 
-
 _UNITS = ["B", "KiB", "MiB", "GiB", "TiB", "PiB"]
+
 
 class ProfilerState(Enum):
     """Enumeration of profiler states"""
     STOPPED = "stopped"
     STARTED = "started"
 
+
 class Profiler:
     """Class to profile the execution of experiments."""
 
-    def __init__(self, expid:str):
+    def __init__(self, expid: str):
         self._profiler = cProfile.Profile()
         self._expid = expid
 
@@ -103,17 +102,16 @@ class Profiler:
             raise AutosubmitCritical(
                 f'Directory {report_path} not writable. Please check permissions.', 7012)
 
-
         stream = io.StringIO()
         date_time = datetime.now().strftime('%Y%m%d-%H%M%S')
 
         # Generate function-by-function profiling results
         sort_by = SortKey.CUMULATIVE
         stats = pstats.Stats(self._profiler, stream=stream)  # generate statistics
-        stats.strip_dirs().sort_stats(sort_by).print_stats() # format and save in the stream
+        stats.strip_dirs().sort_stats(sort_by).print_stats()  # format and save in the stream
 
         # Generate memory profiling results
-        mem_total : float = self._mem_final - self._mem_init  # memory in Bytes
+        mem_total: float = self._mem_final - self._mem_init  # memory in Bytes
         unit = 0
         # reduces the value to its most suitable unit
         while mem_total >= 1024 and unit <= len(_UNITS) - 1:
@@ -128,16 +126,17 @@ class Profiler:
             _generate_title("Memory Profiling"),
             f"MEMORY CONSUMPTION: {mem_total} {_UNITS[unit]}.",
             ""
-        ]).replace('{', '{{').replace('}', '}}') # escape {} so Log can call str.format
+        ]).replace('{', '{{').replace('}', '}}')  # escape {} so Log can call str.format
 
         Log.info(report)
 
         stats.dump_stats(Path(report_path, f"{self._expid}_profile_{date_time}.prof"))
         with open(Path(report_path, f"{self._expid}_profile_{date_time}.txt"),
-            'w', encoding='UTF-8') as report_file:
+                  'w', encoding='UTF-8') as report_file:
             report_file.write(report)
 
         Log.info(f"[INFO] You can also find report and prof files at {report_path}\n")
+
 
 def _generate_title(title="") -> str:
     """
@@ -153,6 +152,7 @@ def _generate_title(title="") -> str:
     message = title.center(max_len)
     return "\n".join([separator, message, separator])
 
+
 def _get_current_memory() -> int:
     """
     Return the current memory consumption of the process in Bytes.
@@ -160,4 +160,4 @@ def _get_current_memory() -> int:
     :return: The current memory used by the process (Bytes).
     :rtype: int
     """
-    return psutil.Process(os.getpid()).memory_info().rss
+    return Process(os.getpid()).memory_info().rss

--- a/test/integration/test_run_command_integration.py
+++ b/test/integration/test_run_command_integration.py
@@ -18,13 +18,11 @@
 import os
 import pwd
 import sqlite3
-from contextlib import suppress
 from pathlib import Path
 from textwrap import dedent
 from typing import Any
 
 import pytest
-from psutil import ZombieProcess
 
 _EXPID = 't000'
 """The experiment ID used throughout the test."""
@@ -463,17 +461,14 @@ def test_run_interrupted(
     _assert_exit_code(final_status, exit_code)
 
     current_statuses = 'SUBMITTED, QUEUING, RUNNING'
-    # TODO: When you call ``.cmdline`` in a ``psutil.Process`` during an iteration,
-    #       if it is a zombie it might raise a ``ZombieProcess`` error.
-    with suppress(ZombieProcess):
-        as_exp.autosubmit.stop(
-            all_expids=False,
-            cancel=False,
-            current_status=current_statuses,
-            expids=_EXPID,
-            force=True,
-            force_all=True,
-            status='FAILED')
+    as_exp.autosubmit.stop(
+        all_expids=False,
+        cancel=False,
+        current_status=current_statuses,
+        expids=_EXPID,
+        force=True,
+        force_all=True,
+        status='FAILED')
 
     exit_code = as_exp.autosubmit.run_experiment(expid=_EXPID)
     _assert_exit_code(final_status, exit_code)

--- a/test/unit/helpers/test_processes.py
+++ b/test/unit/helpers/test_processes.py
@@ -16,6 +16,7 @@
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
 
 from getpass import getuser
+from psutil import ZombieProcess
 from typing import List, Optional
 
 import pytest
@@ -72,6 +73,26 @@ def test_retrieve_expids(mocker, expids: List[str], expected_retrieved: int, use
     assert len(found) == expected_retrieved
 
 
+def test_retrieve_expids_but_listing_raises_zombies(mocker):
+    mocker.patch('autosubmit.helpers.processes.process_iter', side_effect=ZombieProcess(1984))
+    expids = retrieve_expids()
+    assert type(expids) is list and len(expids) == 0
+
+
+def test_retrieve_expids_but_cmdline_raises_zombies(mocker):
+    a000 = _create_process(mocker, 'a000', getuser())
+    a001 = _create_process(mocker, 'a001', getuser())
+    a001.cmdline.side_effect = ZombieProcess(1984)
+    processes = [
+        a000,
+        a001
+    ]
+    mocker.patch('autosubmit.helpers.processes.process_iter', return_value=processes)
+    expids = retrieve_expids()
+    assert len(expids) == 1
+    assert expids[0] in a000.cmdline()
+
+
 @pytest.mark.parametrize(
     'expid,command,pid',
     [
@@ -111,3 +132,12 @@ def test_process_id_multiple_found(mocker):
 
     assert pid_found is 1
     assert mocked_log.warning.call_count == 1
+
+
+def test_process_id_but_there_are_zombies(mocker):
+    """Test that if the listing of process crashes due to zombies, we get an empty list."""
+    mocker.patch('autosubmit.helpers.processes.process_iter', side_effect=ZombieProcess(1984))
+
+    pid_found = process_id('a000')
+
+    assert pid_found is None

